### PR TITLE
feat(environment-variable): file as environment variable

### PR DESCRIPTION
**Pull request by bdebon**
# What does this PR do?

[> Link to the JIRA ticket](https://qovery.atlassian.net/jira/software/projects/FRT/boards/23?selectedIssue=FRT-637)

- Integrate "Mount file as secret" feature

CleanShot 2023-02-08 at 18 12 07@2x

---

## PR Checklist

- [ ] This PR introduces breaking change(s) and has been labeled as such
- [ ] This PR introduces new store changes
- [x] I have followed the library pattern i.e `feature`, `ui`, `data`, `utils`
- [x] I made sure the code is type safe (no any)

### CODE REVIEW

**Comment by evoxmusic** [comment]
A preview environment was automatically created via Qovery.
Click on the link below to follow its deployment and use it.
:point_right: [[PR] staging - feat(environment-variable): file as environment variable - 2023-02-08T17:12:29Z](https://new.console.qovery.com/organization/141c07c8-0dd9-4623-983b-3fdd61867255/project/977bd069-20f7-4b24-bd95-28b44174cf43/environment/4e6f179b-a363-4037-a6c2-77bf401e6657/services/general)

 Another comment will be posted when deployments are terminated

**Comment by evoxmusic** [comment]
Your preview environment has been successfully deployed !
Click on the link below to open your service:
:point_right: [console](https://z4e6f179b-z290d9099-gtw.za8ad0659.bool.sh)
:point_right: [storybook](https://z4e6f179b-z3c8783fa-gtw.za8ad0659.bool.sh)

File: libs/domains/organization/src/lib/slices/cluster/cluster.actions.ts [hunk 1]
```diff
@@ -9,7 +9,7 @@ export const postClusterActionsUpdate = createAsyncThunk<any, { organizationId:
   'clusterActions/update',
   async (data, { dispatch }) => {
     try {
-      const response = await clusterApi.updateCluster(data.organizationId, data.clusterId)
+      const response = await clusterApi.editCluster(data.organizationId, data.clusterId)
```
Comment (bdebon) [inline libs/domains/organization/src/lib/slices/cluster/cluster.actions.ts:12]: There was an update in the api doc, I had to change that.

File: libs/shared/ui/src/lib/components/icon/icon-awesome.enum.tsx [hunk 2]
```diff
@@ -92,5 +92,8 @@ export enum IconAwesomeEnum {
   FILTER = 'icon-solid-filter',
   CLOUD_WORD = 'icon-solid-cloud-word',
   PUZZLE_PIECE = 'icon-solid-puzzle-piece',
+  FILE_LOCK = 'icon-solid-file-lock',
```
Comment (bdebon) [inline libs/shared/ui/src/lib/components/icon/icon-awesome.enum.tsx:95]: this one is missing for now, I'm waiting for Yael's font

File: package.json [hunk 3]
```diff
@@ -54,7 +54,7 @@
     "msw": "^0.42.1",
     "postcss": "8.4.16",
     "posthog-js": "^1.24.0",
-    "qovery-typescript-axios": "^1.1.85",
+    "qovery-typescript-axios": "^1.1.100",
```
Comment (bdebon) [inline package.json:57]: 🎉

File: libs/pages/application/src/lib/feature/crud-environment-variable-modal-feature/crud-environment-variable-modal-feature.tsx [hunk 4]
```diff
@@ -51,6 +53,7 @@ export function CrudEnvironmentVariableModalFeature(props: CrudEnvironmentVariab
   const { enableAlertClickOutside } = useModal()
 
   useEffect(() => {
+    console.log(props.variable)
```
Comment (RemiBonnet) [inline libs/pages/application/src/lib/feature/crud-environment-variable-modal-feature/crud-environment-variable-modal-feature.tsx:56]: log

File: libs/pages/application/src/lib/feature/page-variables-feature/page-variables-feature.tsx [hunk 5]
```diff
@@ -30,7 +30,9 @@ export function PageVariablesFeature() {
   useDocumentTitle('Environment Variables – Qovery')
   const dispatch = useDispatch<AppDispatch>()
   const { applicationId = '' } = useParams()
-  const [placeholder] = useState(environmentVariableFactoryMock(5))
+  const [placeholder] = useState<EnvironmentVariableSecretOrPublic[]>(
```
Comment (RemiBonnet) [inline libs/pages/application/src/lib/feature/page-variables-feature/page-variables-feature.tsx:33]: Seems to be a simple variable if you don't use the `setState` no?

File: libs/pages/application/src/lib/ui/crud-environment-variable-modal/crud-environment-variable-modal.tsx [hunk 6]
```diff
@@ -57,6 +59,31 @@ export function CrudEnvironmentVariableModal(props: CrudEnvironmentVariableModal
           />
         )}
 
+        {props.isFile &&
+          (props.type === EnvironmentVariableType.ALIAS ||
+          props.type === EnvironmentVariableType.OVERRIDE ||
+          props.mode === EnvironmentVariableCrudMode.EDITION ? (
+            <InputText className="mb-3" name="parent value" value={getValues().mountPath} label="Path" disabled />
```
Comment (RemiBonnet) [inline libs/pages/application/src/lib/ui/crud-environment-variable-modal/crud-environment-variable-modal.tsx:66]: Do you have an typo for the name  `name="parent value"` ?

File: libs/shared/ui/src/lib/components/buttons/button-action/button-action.tsx [hunk 7]
```diff
@@ -22,6 +22,7 @@ export interface ButtonActionProps {
   className?: string
   onClick?: () => void
   menus?: MenuData
+  menuMain?: MenuData
```
Comment (RemiBonnet) [inline libs/shared/ui/src/lib/components/buttons/button-action/button-action.tsx:25]: Maybe it should be better to rename it with "dropdown" ?

File: libs/shared/utils/src/lib/tools/environment-variable-file.ts [hunk 8]
```diff
@@ -0,0 +1,48 @@
+import { APIVariableTypeEnum } from 'qovery-typescript-axios'
+import { EnvironmentVariableSecretOrPublic } from '@qovery/shared/interfaces'
+
+// todo remove the any once the api is updated
```
Comment (RemiBonnet) [inline libs/shared/utils/src/lib/tools/environment-variable-file.ts:4]: The api isn't updated now ?

File: libs/shared/utils/src/lib/tools/environment-variable-file.ts [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+): string => {
+  if (!variable) return ''
+  if (variable.variable_type === APIVariableTypeEnum.ALIAS) {
+    if (variable.aliased_variable?.variable_type === APIVariableTypeEnum.FILE) {
+      return variable.aliased_variable.mount_path
+    }
+
+    if (variable.aliased_secret?.variable_type === APIVariableTypeEnum.FILE) {
+      return variable.aliased_secret.mount_path
+    }
```
Comment (RemiBonnet) [inline libs/shared/utils/src/lib/tools/environment-variable-file.ts:36]: This function is quite difficult to read (lots of if) maybe you can add more spacing or small comments ?

**Review by RemiBonnet — COMMENTED** [review]
Well @bdebon thanks !! 💯 my feedback: 

- Can you allow the click on the "Secret Variable" ? (I think you don't use the props "title" of this component. BTW, it may be too close with the top
Capture d’écran 2023-02-13 à 17 43 35

- When I created an secret file variable the value is null, but works when I reload and we don't see the left label "file. And when I updated my variable I see a flash with the "null" value
Capture d’écran 2023-02-13 à 17 45 17

File: libs/pages/application/src/lib/feature/page-variables-feature/page-variables-feature.tsx (hunk 5 above)
Comment (bdebon) [reply libs/pages/application/src/lib/feature/page-variables-feature/page-variables-feature.tsx:33]: I use this placeholder in a useEffect below and as it is generated by a factoryMock, the value of placeholder would change at every render and then retrigger the useEffect in loop...

**Comment by bdebon** [comment]
The blink and the null behavior on secret is a problem backend side, Benjamin Chastanier is on it. Good catch Remi!

File: libs/pages/application/src/lib/feature/page-variables-feature/page-variables-feature.tsx (hunk 5 above)
Comment (RemiBonnet) [reply libs/pages/application/src/lib/feature/page-variables-feature/page-variables-feature.tsx:33]: useMemo can be a solution to avoid the render ?

**Review by RemiBonnet — APPROVED** [review]

**Pull request merged**